### PR TITLE
Several cleanups

### PR DESCRIPTION
**Pull request by juliusv**
Several smaller improvements and cleanups (see individual commits).

These have been done without having any collectd to test against.
I'll try to set up a collectd soon and add tests.

### CODE REVIEW

**Comment by juliusv** [comment]
@brian-brazil

File: main.go [hunk 1]
```diff
@@ -161,8 +183,8 @@ func (c *CollectdCollector) processSamples() {
 	}
 }
 
-// Implements Collector.
-func (c CollectdCollector) Collect(ch chan<- prometheus.Metric) {
+// Collect implements prometheus.Collector.
+func (c collectdCollector) Collect(ch chan<- prometheus.Metric) {
 	ch <- lastPush
 	c.mu.Lock()
 	samples := c.samples
```
Comment (juliusv) [inline main.go]: This only creates a copy of the map metadata, but that's not enough to protect a reader against changes in the map. I think we'll either need to lock the mutex for the entire metrics collection period, or create a deep copy under the mutex upfront.

Comment (juliusv) [inline main.go]: Minimum test case to illustrate this:

```golang
package main

import "sync"

func main() {
  m := map[int]int{}
  mu := sync.Mutex{}

  go func() {
    for {
      mu.Lock()
      m[0] = 1
      delete(m, 0)
      mu.Unlock()
    }
  }()

  for {
    mu.Lock()
    s := m
    mu.Unlock()

    for _ = range s {
    }
  }
}
```

```
$ GOMAXPROCS=2 go run -race maps.go
==================
WARNING: DATA RACE
Read by main goroutine:
  runtime.mapiterinit()
      /usr/local/go/src/pkg/runtime/hashmap.goc:1011 +0x0
  main.main()
      /home/julius/tmp/maps.go:23 +0x1a7

Previous write by goroutine 4:
  runtime.mapassign1()
      /usr/local/go/src/pkg/runtime/hashmap.goc:925 +0x0
  main.func·001()
      /home/julius/tmp/maps.go:12 +0x94

Goroutine 4 (running) created at:
  main.main()
      /home/julius/tmp/maps.go:16 +0x142
==================
```

Comment (juliusv) [inline main.go]: I pushed two more commits, one fixing this.

File: main.go [hunk 2]
```diff
@@ -29,8 +42,8 @@ var (
 	addr     = flag.String("listen-address", ":6060", "The address to listen on for HTTP requests.")
 	lastPush = prometheus.NewGauge(
 		prometheus.GaugeOpts{
-			Name: "collectd_last_push",
-			Help: "Unixtime the collectd exporter was last pushed to.",
+			Name: "collectd_last_push_timestamp_seconds",
```
Comment (brian-brazil) [inline main.go]: _unixtime?

seconds is a duration, not a timestamp.

File: main.go [hunk 3]
```diff
@@ -114,23 +136,23 @@ func (c *CollectdCollector) collectdPost(w http.ResponseWriter, r *http.Request)
...
 			labels["instance"] = metric.Host
 			c.ch <- &collectdSample{
-				Name:    name,
+				Name:    metricName(metric, metric.Dstypes[i], metric.Dsnames[i]),
 				Labels:  labels,
-				Help:    help,
+				Help:    metricHelp(metric, metric.Dstypes[i], metric.Dsnames[i]),
 				Value:   value,
-				Gauge:   metric.Dstypes[i] != "counter",
+				Type:    metricValue(metric.Dstypes[i]),
```
Comment (brian-brazil) [inline main.go]: metricType would be clearer

File: main.go [hunk 4]
```diff
@@ -161,39 +183,32 @@ func (c *CollectdCollector) processSamples() {
...
-				prometheus.CounterOpts{
-					Name:        sample.Name,
-					Help:        sample.Help,
-					ConstLabels: sample.Labels})
-			counter.Set(sample.Value)
-			ch <- counter
-		}
+		ch <- prometheus.MustNewConstMetric(
+			prometheus.NewDesc(sample.Name, sample.Help, []string{}, sample.Labels),
+			prometheus.GaugeValue,
```
Comment (brian-brazil) [inline main.go]: sample.Type

File: main.go (hunk 3 above)
Comment (juliusv) [inline main.go]: Ah, that's what happens if I code too late. Changed.

File: main.go (hunk 4 above)
Comment (juliusv) [inline main.go]: Fixed, thanks.

File: main.go (hunk 2 above)
Comment (juliusv) [inline main.go]: We use "_timestamp_seconds" for Unix timestamps in several places (at least internally). "seconds" as a unit vs. nanoseconds or microseconds, and "timestamp" to indicate that it's a unix timestamp in general. Happy to discuss this later, but will merge for now, as the benefits of getting the bug fixes in are probably good given that it's now linked from Hacker News :)

**Pull request merged**